src/puppetlabs/puppetdb/scf/migrate.clj

```diff
@@ -1374,21 +1374,32 @@
         WHERE f.value IS NULL AND fv.value IS NULL")
 
   (log/info (trs "[5/8] Cleaning up duplicate null values..."))
-  (let [json-null-value-id (some-> (jdbc/query-to-vec "select id from fact_values
-                                                       where value_type_id = 5 and value = 'null'")
-                                   first
-                                   :id)
-        real-null-value-id (some-> (jdbc/query-to-vec "select id from fact_values
-                                                       where value_type_id = 4 and value is null")
-                                   first
-                                   :id)]
-    (if (and json-null-value-id real-null-value-id)
-      (do
-        (jdbc/do-prepared "update facts_transform set fact_value_id = (?) where fact_value_id = (?)"
-                          [real-null-value-id json-null-value-id])
-        (jdbc/do-prepared "delete from fact_values where id = (?)"
-                          [json-null-value-id]))
-      (log/info (trs "  ... none found"))))
+  ;; only do this if the DB has some null values
+  (when (pos? (-> (jdbc/query-to-vec "select count(*) from fact_values
+                                      where value_type_id in (4,5)
+                                      and value in (null, 'null')")
+                  first
+                  :count))
+   (let [real-null-value-id (let [existing-id (some-> (jdbc/query-to-vec
+                                                       "select id from fact_values
+                                                       where value_type_id=4 and value is null")
+                                                      first
+                                                      :id)]
+                              (or existing-id
+                                  (-> (jdbc/query-to-vec
+                                       "insert into fact_values (value_type_id, value) values (4, null) returning id")
+                                      first
+                                      :id)))
```

Comment:
seems like existing-id/real-null-value-id would be nicer as two separate bindings in the top-level let

Reply:
I believe this will violate an fk constraint if the install has never seen a null-valued fact

Reply:
ah nvm, last comment is handled

Reply:
actually, it isn't. If you get into this code because you have a string-valued null in a json type, I think this will still violate a constraint

Reply:
forget the last three comments -- the value types are hardcoded in the migration